Jupyter server

oliver-sanders opened the pull request:
closes #219, #218, #45
unblocks #137
addresses #171 
siblings: https://github.com/cylc/cylc-flow/pull/4331, https://github.com/cylc/cylc-ui/pull/711

I think this also closes (testing required to confirm):
* #103 (https://github.com/cylc/cylc-uiserver/pull/230#issuecomment-891078051)
* #215

### Summary & Highlights

* :rocket: Jupter Server replaces our bespoke Tornado server.
* :smile: Nice simplification of UIS codebase.
* :gift: This means we get lots of lovely Jupyter goodies and tighter integration into Jupyter land.
* :muscle: Jupyter servers can run multiple Jupyter Server extensions (e.g. you can run the Cylc UIS and Lab on the same server).
* :hugs: This opens up the possibility of tighter integration between the Cylc UIS and Juoyter Lab.
* :people_holding_hands: We now have two modes, single and multi user.
  * :person_curly_hair: single-user uses token authentication, no authorisation, no need to run the hub.
  * :people_holding_hands: multi-user functionality provides authorisation, requires the hub.
* :abacus: There is the potential to wrap the single-user mode a a native-like application using electron or similar.

### CLI

Everything can be run using the `jupyter` commands:

```bash
# launch standalone UIS (single-user mode)
# [automatically sources Cylc configuration files]
jupyter-cylc

# launch the hub (multi-user mode)
# [you must point hub at the Cylc configuration files or edit your user/site Jupyter config]
jupyterhub

# the command the hub spawns
# [automatically sources Cylc configuration files]
jupyter-cylchubapp
```

Existing jupyter server installations can be configured to run the Cylc UIS extensions, example config:

```python
c.ServerApp.jpserver_extensions = {
        "cylc.uiserver": True,
}
```

For tighter integration with Cylc, improved usability for those unfamiliar with the Jupyter ecosystem and command whitelisting purposes there are Cylc equivalents to these commands:

```bash
# all Cylc commands source the cylc configurations out of the box

# launch single-user
cylc gui

# launch multi-user
cylc hub

# command the hub spawns
cylc hubapp
```

### Caveats

* ~https://github.com/cylc/cylc-uiserver/issues/231~
  [edit] now resolved!

  Issue to do with redirects from static endpoint (e.g. the UI) to the hub which are largely broken on this branch at the moment.

  This has the side-effect that it makes it look like you can access UI Servers you are not authorised to simply because the web application loads without any login redirect or authorisation warning. However, the data endpoints are all protected.

* ~https://github.com/jupyter-server/jupyter_server/issues/566~
  [edit] now partially resolved, if the user is a string but not "anonymous" then it is a token.

  We are pinned to `jupyter_server==1.10.1` until this issue is resolved as in order to determine when a connection is token authenticated we are currently relying on a bug which has since been fixed.

### Follow On

* https://github.com/cylc/release-actions/issues/17
  (required in order to release this package, high priority).
* ~https://github.com/cylc/cylc-uiserver/issues/231~
  (required for correct authentication redirection on Cylc endpoints, high priority).
* https://github.com/cylc/cylc-uiserver/issues/235
  (requires `jupyter_server>=1.10.2`, low priority).

### Misc

Ditch the `setup.py` file, we can do everything we need using `setup.cfg` which is much nicer :tada:, no more arbitrary code execution! We should copy this change to our other repos in due course.

Also adopted the new(ish) standard `pyproject.toml` file.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] Docs: https://github.com/cylc/cylc-doc/pull/285
- [ ] Created an issue at [cylc-uiserver conda-forge repository](https://github.com/conda-forge/cylc-uiserver-feedstock) with version changes (if you changed dependencies in `setup.py`, see `recipe/meta.yaml`).

kinow commented on CHANGES.md:
```diff
@@ -10,6 +10,17 @@ creating a new release entry be sure to copy & paste the span tag with the
+-------------------------------------------------------------------------------
+## __cylc-uiserver-0.6.0 (<span actions:bind='release-date'>Released 2021-??-??</span>)__
+
+Multi-user functionality implented.
```
s/implented/implemented

kinow replied:
:point_up: @oliver-sanders

oliver-sanders commented on conftest.py:
```diff
@@ -0,0 +1,9 @@
+pytest_plugins = [
+    'jupyter_server.pytest_plugin'
```
This pulls in the `jupyter_server` testing framework.

oliver-sanders commented on cylc/uiserver/app.py:
```diff
@@ -0,0 +1,364 @@
+from cylc.uiserver.workflows_mgr import WorkflowsManager
+
+
+class MyExtensionHandler(JupyterHandler):
+
+    @tornado.web.authenticated
+    def get(self):
+        self.write('hello cylc')
+
+    @tornado.web.authenticated
+    def post(self):
+        ...
```
Temporary handler useful for debugging.

oliver-sanders commented on cylc/uiserver/app.py:
```diff
@@ -0,0 +1,364 @@
+    examples = '''
+        cylc gui    # start the cylc GUI
+    '''
+    config_file_paths = list(
+        map(
+            str,
+            [
+                # base configuration - always used
+                Path(uis_pkg).parent,
+                # site configuration
+                Path('/etc/cylc/hub'),
+                # user configuration
+                Path('~/.cylc/hub').expanduser()
+            ]
+        )
+    )
```
This registers the Cylc configuration files with Jupyter Server, they get loaded automatically.

datamel replied:
Does this need to support `CYLC_SITE_CONF_PATH`?

oliver-sanders commented on cylc/uiserver/handlers.py:
```diff
@@ -13,109 +13,153 @@
 
+    https://github.com/jupyter-server/jupyter_server/pull/562
+    """
+    if isinstance(user, bytes):
+        # Cookie authentication:
+        # * The URL token is added to a secure cookie, it can then be
+        #   removed rrom the URL for subsequent requests, the cookie is
+        #   used in its place.
+        # * If the token was used token_authenticated is True.
+        # * If the cookie was used it is False (despite the cookie auth
+        #   being derived from token auth).
+        # * Due to a bug in jupyter_server the user is returned as bytes
+        #   when cookie auth is used so at present we can use this to
+        #   tell.
+        #   https://github.com/jupyter-server/jupyter_server/pull/562
+        # TODO: this hack is obviously not suitable for production!
+        return True
```
:eyes:

oliver-sanders commented on cylc/uiserver/handlers.py:
```diff
@@ -13,109 +13,153 @@
-class BaseHandler(HubOAuthenticated, web.RequestHandler):
 
-    def set_default_headers(self) -> None:
-        self.set_header("X-JupyterHub-Version", jupyterhub_version)
-        self.set_header("Access-Control-Allow-Origin", "*")
-        self.set_header("Access-Control-Allow-Headers", "x-requested-with")
-        self.set_header('Access-Control-Allow-Methods', 'POST, GET, OPTIONS')
```
Jupyter Server does all this stuff for us :tada:

oliver-sanders commented on cylc/uiserver/handlers.py:
```diff
@@ -13,109 +13,153 @@
-    # allow_admin = True
+    @property
+    def hub_users(self):
+        # allow all users (handled by Cylc authorisation decorator)
+        return None
 
-    def initialize(self, path):
-        self._static = path
+    @property
+    def hub_groups(self):
+        # allow all groups (handled by Cylc authorisation decorator)
+        return None
```
This effectively deactivates the default Jupyter Server authorisation allowing us to open up these endpoints. All secure endpoints MUST have `@authorised` decorators.

oliver-sanders commented on cylc/uiserver/hubapp.py:
```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2015 Project Jupyter Contributors
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice,
+#    this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
```
DO NOT MERGE UNTIL RESOLVED

We will need to evaluate the licencing implications of this which will likely mandate re-writing the code.

oliver-sanders replied:
Resolved.

oliver-sanders commented on cylc/uiserver/tests/test_uiserver.py:
```diff
@@ -18,101 +18,77 @@
-    uis = CylcUIServer(8000, '/prefix/', ui_build_dir / '1.0')
-    # argument wins
-    assert uis.ui_path == ui_build_dir / '1.0'
+# def test_ui_build_dir_argument(mock_config, ui_build_dir):
+#     """The ui_build_dir argument should override any config value."""
+#     # config asks for 2.0
+#     mock_config(CylcUIServer={
+#         'ui_build_dir': ui_build_dir / '2.0'
+#         })
+#     uis = CylcUIServer()
+#     uis.initialize_settings()
+#     # normally this would be used
+#     assert uis.ui_path == ui_build_dir / '2.0'
+#     # argument asks for 1.0
+#     uis = CylcUIServer(8000, '/prefix/', ui_build_dir / '1.0')
+#     uis.initialize_settings()
+#     # argument wins
+#     assert uis.ui_path == ui_build_dir / '1.0'
```
Build dir argument has been removed, you can still use environment variables + logic in your `jupyter_config` file.

oliver-sanders replied:
Update, CLI args for jupyter_server extensions were broken but have since been fixed in version 1.10.2, however, we can't use that yet due to the issue of detecting token authenticated connections (see caveats in issue description above).

oliver-sanders replied:
Bumped to https://github.com/cylc/cylc-uiserver/issues/235

oliver-sanders commented:
(deleted more than I meant to in the last commit)

oliver-sanders commented:
I think this closes #103 as we are now using `make_singleuser_app`:

https://github.com/jupyterhub/jupyterhub/blob/c00c3fa28703669b932eb84549654238ff8995dc/jupyterhub/singleuser/mixins.py#L300-L310

@kinow

kinow commented:
> I think this closes #103 as we are now using `make_singleuser_app`:
> 
> https://github.com/jupyterhub/jupyterhub/blob/c00c3fa28703669b932eb84549654238ff8995dc/jupyterhub/singleuser/mixins.py#L300-L310
> 
> @kinow, any thoughts?

Looks like it'll fix the SSL issue! :clap: great work @oliver-sanders !

kinow commented on cylc/uiserver/handlers.py:
```diff
@@ -13,109 +13,153 @@
+            # is provided.
+            raise web.HTTPError(403, reason='Forbidden')
+        handler.log.critical(
+            f'authorise {user} '  # type: ignore
```
Is it dangerous to log the User's token?

oliver-sanders replied:
Almost certainly, apologies this is some debug kicking around due to the problem of detecting token-authenticated sessions. Hopefully there will be a solution there in the coming days :crossed_fingers:.

kinow reviewed (commented):
Did a quick test, had to update `pip install -U pip` to be able to `pip install -e .`. PEP 517 looks interesting! Had seen that toml file in other projects, but never used myself. Had a quick look at the new files, syntax looks simple :+1: 

## Single user

After installing it did a quick `jupyter-cylc`, and was surprised to see the UI "kinda" working from `master` :smile:  some console errors, and the hierarchical scan was not working.

Realized it's because it's using the 0.5.0 release of the UI. Is there a way to run `cylc gui` or `jupyter cylc` pointing to the development dist folder of Cylc UI? That would be a faster way to test the UI without the mocked data I think.

## Jupyter Hub

`cylc hub`, and that's it :tada: 

It's almost 4PM Friday, so I will stop reviewing it, and instead start Monday morning since I think I should be able to do a much better job with fresh :coffee: :smile: 

But impressive work! :tada:

oliver-sanders commented:
> Is there a way to run cylc gui or jupyter cylc pointing to the development dist folder of Cylc UI?

Yep, set the `build_dir` traitlet to point at your `build_dir`, some basic details in the `README`:

https://github.com/cylc/cylc-uiserver/pull/230/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R215-R223

Some more info in the traitlets docstrings (we can auto-generate HTML docs from these later).

The way I do it is using a `CYLC_DEV` environment variable which I detect in my `jupyter_config`:

```python
import os
from pathlib import Path

if os.environ.get('CYLC_DEV', 'false').lower() == 'true':
    print('### Using Dev UI')  # for debug purposes

    ## pick up the development UI build

    # 0.5.0
    # c.UIServer.ui_build_dir = Path('~/cylc-ui/dist').expanduser()
    # oliver-sanders/jupyter-server
    c.CylcUIServer.ui_build_dir = Path('~/cylc-ui/dist').expanduser()

    ## pass the CYLC_DEV env var through to the UIS if spawned via jupyter_hub

    # 0.5.0
    # c.Spawner.cmd = ['env', 'CYLC_DEV=true', 'cylc', 'uis']
    # oliver-sanders/jupyter-server
    c.Spawner.cmd = ['env', 'CYLC_DEV=true', 'cylc', 'hubapp']
```

We also use our `jupyter_config` for creating a self-signed SSL cert so we can use HTTPS automatically.

kinow commented:
I will have another read of the readme and comments here on Monday. I have some learning curve (I also want to read more about Jupyter Server) but so far really happy with the improvements for the UIS! :tada: Thanks @oliver-sanders !

oliver-sanders commented:
Restored `CYLC_SITE_CONF_PATH` functionality, great spot @datamel.

datamel reviewed (commented):
A couple of things so far, still have some code reviewing and check multi user access is blocked. 

Related to our conversation earlier, so you are aware but functionality is affected...
On testing the graphiql basic query: 
```
query {workflows { id 
  status }
}

```
I get a 

```
{
  "message": "Response.text: Body has already been consumed.",
  "stack": "b/</<@http://localhost:8888/cylc/js/chunk-42973249.8f733a85.js:1:467562\n"
}
```
with a server log of
```
[W 2021-08-13 12:44:30.859 ServerApp] 403 POST /cylc/graphql (127.0.0.1): '_xsrf' argument missing from POST
[W 2021-08-13 12:44:30.860 ServerApp] 403 POST /cylc/graphql (127.0.0.1) 1.87ms referer=http://localhost:8888/cylc/?token=blahblahblah

``` 

Also, on running a workflow I had an error which I have not seen before (not sure if related to this branch for sure - possibly/probably not but thought I'd post it in case)...

![image](https://user-images.githubusercontent.com/37735232/129350593-23e63c83-81e7-4bfb-a1cd-f6c3597d6ba8.png)

kinow commented on README.md:
```diff
@@ -1,22 +1,21 @@
+This project contains the Cylc UI Server which provides the Cylc GUI
 used to serve the Cylc UI, and to communicate with running Cylc Schedulers.
 
 [Cylc Website](https://cylc.org/) |
```
Ping @hjoliver

oliver-sanders replied:
> https://cylc.org/ currently forwards to https://niwa.co.nz/

Oh, that's changed recently it used to redirect to GH pages, yikes!

> Response.text: Body has already been consumed.

Hmmm, I've seen this error before, I think my rebase of the cylc-ui branch might not have been fully successful.

kinow commented on README.md:
```diff
@@ -25,24 +24,37 @@ interface.
 
   This is a web server which serves the Cylc web UI. It connects to running
   workflows and workflow databases to provide the information the UI displays.
+  It is a [Jupyter Server](https://github.com/jupyter-server/jupyter_server).
```
A Jupyter Server application, or a Jupyter web Application? I'm reading their docs, from what I understood, Jupyter Server is the utility code for creating applications or extensions. Applications being complete web applications that use the Jupyter Server, and Extensions only extend the REST API.

kinow replied:
Huh, so looks like everything is an extension, but there's a special `ExtensionApp` that makes an extension that behaves like the old applications  (like notebook, and current UI Server I think?). So everything is an extension? :smile:

oliver-sanders replied:
So there's some similar but different terminology here which makes things confusing.

The Cylc UI Server (as of this PR) is a "Jupyter Server extension applicaiton".

* The "extension" part is because it is extending Jupyter Server.
* The application part is because it is a configurable traitlets application.

Extension applications can:

* Add static files (automatically served by the server)
  * Not doing this right now because of https://github.com/cylc/cylc-uiserver/issues/231
* Add handlers.
* Request resources from the server.

Jupyter server also offers non-configurable extensions (ones which aren't applications), these are cruder.

kinow commented on README.md:
```diff
@@ -52,13 +64,56 @@ $ conda install cylc-uiserver
+
+The Cylc UIServer is a
+[Jupyter Server](https://github.com/jupyter-server/jupyter_server)
+extension (like [JupyterLab](https://github.com/jupyterlab/jupyterlab)).
```
Ah, I guess it's an extension then, I thought it was an application :+1:

oliver-sanders replied:
If splitting hairs it is technically an "extension application"!

From the users perspective the difference between an "extension" and an "extension application" is invisible.

kinow replied:
Brilliant, thanks @oliver-sanders ! It was more so I knew the correct term in case I needed to report an issue upstream, or talk to someone else about Jupyter Server. But extension application is quite a mouthful, I'll probably just use jupyter server, or jupyter server app, or whichever my :brain: picks first.

kinow commented on cylc/uiserver/app.py:
```diff
@@ -0,0 +1,369 @@
+    Cylc - A user interface for monitoring and controlling Cylc workflows.
+    '''
+    examples = '''
+        cylc gui    # start the cylc GUI
```
Note: I think there's a NBSP character here. Even though GH UI doesn't show it, my IDE distinguishes spaces and NBSP's.

![image](https://user-images.githubusercontent.com/304786/129639880-f98f7e16-8d96-4ee2-b306-b11c959ec818.png)

Harmless, but may break some tools/editors (I think moving cursor over the space and pressing `ga` in vim shows the hexcode/etc to compare with space)

oliver-sanders replied:
Harmless but can be annoying, it's what happens when you press alt+space on Mac OS.

kinow commented on cylc/uiserver/app.py:
```diff
@@ -0,0 +1,369 @@
+
+    async def stop_extension(self):
+        for sub in self.data_store_mgr.w_subs.values():
+            sub.stop()
```
Is `stop_extension` stopping the Python executable? I think it is, but asking just in case it's not. If not, I think we need to delete from the dictionary the stopped subs.

Also, should we wrap in a try/catch these calls if terminating program? To make sure we try to stop, shutdown, and destroy?

oliver-sanders replied:
The `stop_extension` hook allows extensions to clean up any resources (e.g. subs) they have created before the server shuts down and the Python process exits.

We need to stop the subs as otherwise the server will hang rather than shutting down.

kinow reviewed (changes requested):
Few typos, some questions, but found no blockers. Once typos fixed, only remaining question I have is whether this should go in first or if @datamel 's #204 should go first. Whichever is merged first, will force the other one to be rebased and I think it will create a few conflicts :+1: 

Great job @oliver-sanders ! There's one file with 75 lines `hubapp.py` that even though it's relatively small, I think it was probably really hard to write. :clap: :clap: :clap:  Thank you!

oliver-sanders commented:
Looks like I've broken GraphiQL, no idea how to fix it, will poke away...

oliver-sanders commented:
Think I've fixed it with the latest change to the cylc-ui PR.

datamel commented:
> Think I've fixed it with the latest change to the cylc-ui PR.

Super, I'll have a go at testing now.

kinow commented:
I think there is one last typo in changes.md.

kinow reviewed (commented):
@oliver-sanders created a new venv, re-installed everything after syncing PR's.

The WebSockets connection is working. As well as the REST call to `userprofile`. But a request to `graphql` to request mutations info is failing I think?

![image](https://user-images.githubusercontent.com/304786/129812012-18ad3aea-db92-4e80-ad14-6440360e5cf8.png)

![image](https://user-images.githubusercontent.com/304786/129812055-764ca81d-8b64-4d08-aa50-d5e4f2a81d84.png)

I used `cylc hub` by the way. The three repos are all pointing to the Jupyter Server Pr's. I changed the version of UI, and it didn't update in the browser, so I think that confirms I am looking at the old UI from 0.5

I used `cylc hub`, and in the comments above

># all Cylc commands source the cylc configurations out of the box

But looking at the logs it shows that it used the `PAMAuthenticator` instead of the `DummyAuthenticator` I've been using, and the wrong location for the UI:

```
[I 2021-08-18 11:30:00.682 CylcHubApp manager:356] cylc.uiserver | extension was successfully linked.
[I 2021-08-18 11:30:00.747 CylcUIServer] Starting Cylc UI Server
[I 2021-08-18 11:30:00.747 CylcUIServer] Serving UI from: /home/kinow/Development/python/workspace/cylc-uiserver/cylc/uiserver/ui/0.5.0

```

Renamed my `~/.cylc/hub/config.py` to `~/.cylc/hub/jupyter_config.py` and it used `DummyAuthenticator`, but still

```
[I 2021-08-18 11:32:49.228 CylcUIServer] Starting Cylc UI Server
[I 2021-08-18 11:32:49.228 CylcUIServer] Serving UI from: /home/kinow/Development/python/workspace/cylc-uiserver/cylc/uiserver/ui/0.5.0
```

But saw this in the logs:

```
[D 2021-08-18 11:32:49.152 CylcUIServer] Looking for jupyter_cylc_config in /home/kinow/.cylc/hub
[D 2021-08-18 11:32:49.153 CylcUIServer] Looking for jupyter_cylc_config in /etc/cylc/hub
[D 2021-08-18 11:32:49.153 CylcUIServer] Looking for jupyter_cylc_config in /home/kinow/Development/python/workspace/cylc-uiserver/cylc/uiserver
```

Tried `ln -s config.py jupyter_cylc_config` in `~/.cylc/hub/config` with no luck.

Finally, reading the README and comparing the key to point to the UI I realized I had the wrong key (I think that changed now? Since yesterday I used that config with the propagate-gscan UI change):

```
# was c.UIServer.ui_build_dir = '~/Development/python/workspace/cylc-ui/dist'
c.CylcUIServer.ui_build_dir = '~/Development/python/workspace/cylc-ui/dist'
```

Now it's showing the version 0.6.0 since I changed that in `package.json`, but the error persists:

![image](https://user-images.githubusercontent.com/304786/129814563-179c2a54-ea6c-4b81-b7af-23ce741af8fb.png)

datamel reviewed (commented):
Still having problems getting the config to load. 

https://github.com/cylc/cylc-uiserver/blob/47096f4518258a00c559f86e34736dccd6f303f2/cylc/uiserver/config.py#L51
I think should be changed to `jupyter_config.py`.

I also unfortunately am getting the same `_xsrf` error on graphiql with all your latest branches checked out.

datamel commented:
Getting 404s on the auto pop-up loading `cylc gui`.

oliver-sanders commented:
Thanks both:

* Mel spotted a mistake which explains why the hub is picking up the wrong config file, sorry.
* Yes, I renamed `UIServer` to `CylcUIServer` as that makes it a bit less ambiguous as a server extension, this means the traitlet namespace has changed too.
* I'm going to try and replicate the xsrf issues seen when running in hub mode.
* I'm going to try and bump us up to the latest Jupyter Server release, you may need to manually (re)install `jupyter_server` at the specified version as `pip install` does not always behave as you would expect it to in this regard.

datamel commented:
> Getting 404s on the auto pop-up loading `cylc gui`.

This issue has self-fixed for me with a re-checkout of all branches, cleaning cookies and re-install. 🤷‍♀️ Not sure what caused it but can't replicate fortunately!

oliver-sanders commented:
Bumped the Jupyter Server version.

> I also unfortunately am getting the same _xsrf error on graphiql with all your latest branches checked out.

>  But a request to graphql to request mutations info is failing I think?

I just cannot reproduce this.

datamel commented:
> * I'm going to try and replicate the xsrf issues seen when running in hub mode.

In case it helps with the replicating, I have found a reliableway to reproduce...
- Run cylc hub, load the ui as normal and run a query, eg
```query{
  workflows {
    id
  }
}
```

This runs as normal, with no _xsrf complaints..

- then go to cylc hub and log out of the ui. Pop up the developer console and look at the network. Log back in.
- You should see 403 for subscriptions and graphql, with output in the ui-server output of 
`403 POST /user/mhall/cylc/graphql (::1): '_xsrf' argument missing from POST` and `Token stored in cookie may have expired
`
As I mentioned in team meet, I think this is not a major problem, I think this problem actually self-corrects.

oliver-sanders commented:
Ok, the last few commits have:

* Responded to outstanding feedback.
* Bumped the Juptyter Server version onto the latest and greatest.
* Fixed #231 (the first caveat) which should resolve the intermittent 403 issues reported earlier.
* Replaced `/cylc/hello` with `/cylc/version` (nice to have a minimal endpoint for debugging).

datamel reviewed (approved):
The problems that I was getting have settled with the last few commits. 
The issue with the 403 after logging out and back in (with the _xsrf error) still occurs but self corrects. I think this is acceptable behaviour as it does not affect functionality.
There is a remaining typo where implanted should be implemented, I think but this should not affect merge, I can fix on the auth pr.
Bravo @oliver-sanders 🥇 - lovely functionality to have!

kinow commented on CHANGES.md:
```diff
@@ -10,6 +10,17 @@ creating a new release entry be sure to copy & paste the span tag with the
+-------------------------------------------------------------------------------
+## __cylc-uiserver-0.6.0 (<span actions:bind='release-date'>Released 2021-??-??</span>)__
+
+Multi-user functionality implanted.
```
s/implanted/implemented

oliver-sanders replied:
:facepalm:

kinow commented on cylc/uiserver/app.py:
```diff
@@ -0,0 +1,343 @@
+from cylc.uiserver.data_store_mgr import DataStoreMgr
+from cylc.uiserver.handlers import (
+    CylcStaticHandler,
+    CylcVersionHandler,
```
These two handlers are not exported in `cylc.uiserver.handlers`. Not too important, we can just ignore here, update the `__all__`, or drop that.

oliver-sanders replied:
Removed the `__all__`.

kinow reviewed (approved):
Looks good to me! Couple comments but I think they are not blockers. Feel free to push the merge button @oliver-sanders (just so you can read the comments before :+1: )

kinow commented:
Two approvals, CI happy, @oliver-sanders resolved all feedback. LGTM :grimacing:
